libparsec/crates/client/src/certif/user_update_profile.rs

```diff
@@ -0,0 +1,179 @@
+    #[error("User not found")]
+    UserNotFound,
+    #[error("Author not allowed")]
+    AuthorNotAllowed,
+    #[error("User revoked")]
+    UserRevoked,
+    #[error("Our clock ({client_timestamp}) and the server's one ({server_timestamp}) are too far apart")]
+    TimestampOutOfBallpark {
+        server_timestamp: DateTime,
+        client_timestamp: DateTime,
+        ballpark_client_early_offset: f64,
+        ballpark_client_late_offset: f64,
+    },
+    #[error("There is no key available in this realm for encryption")]
+    NoKey,
+    #[error(transparent)]
+    InvalidKeysBundle(#[from] Box<InvalidKeysBundleError>),
+    #[error(transparent)]
+    InvalidCertificate(#[from] Box<InvalidCertificateError>),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<ConnectionError> for CertifUpdateUserProfileError {
+    fn from(value: ConnectionError) -> Self {
+        match value {
+            ConnectionError::NoResponse(_) => Self::Offline,
+            // TODO: handle organization expired and user revoked here ?
+            err => Self::Internal(err.into()),
+        }
+    }
+}
+
+impl From<CertifStoreError> for CertifUpdateUserProfileError {
+    fn from(value: CertifStoreError) -> Self {
+        match value {
+            CertifStoreError::Stopped => Self::Stopped,
+            CertifStoreError::Internal(err) => err.into(),
+        }
+    }
+}
+
+pub(super) async fn update_profile(
+    ops: &CertificateOps,
+    user: UserID,
+    new_profile: UserProfile,
+) -> Result<CertificateBasedActionOutcome, CertifUpdateUserProfileError> {
+    if ops.device.user_id == user {
+        // TODO: should an admin be able to downgrade themselves
+        return Err(CertifUpdateUserProfileError::UserIsSelf);
+    }
+
+    // Loop is needed to deal with server requiring greater timestamp
+    let mut timestamp = ops.device.now();
+    loop {
+        let outcome = do_server_command(ops, user, timestamp, new_profile).await?;
+
+        match outcome {
+            DoServerCommandOutcome::Done(outcome) => return Ok(outcome),
+            DoServerCommandOutcome::RequireGreaterTimestamp(strictly_greater_than) => {
+                // TODO: handle `strictly_greater_than` out of the client ballpark by
+                // returning an error
+                timestamp = greater_timestamp(
+                    &ops.device.time_provider,
+                    GreaterTimestampOffset::User,
+                    strictly_greater_than,
+                );
+            }
+        }
+    }
+}
+
+enum DoServerCommandOutcome {
+    Done(CertificateBasedActionOutcome),
+    RequireGreaterTimestamp(DateTime),
+}
+
+async fn do_server_command(
+    ops: &CertificateOps,
+    recipient: UserID,
+    timestamp: DateTime,
+    new_profile: UserProfile,
+) -> Result<DoServerCommandOutcome, CertifUpdateUserProfileError> {
+    // 0) Sanity check to prevent generating and invalid certificate
+
+    if recipient == ops.device.user_id {
```

Comment:
This check has already been done by the caller (i.e. `update_profile`), so two possibilities:

- remove the check from `update_profile`
- consider the check should be done in `update_profile`, in this case the check done here is only for consistency and hence you should do an assert instead of returning an error